File: HoneyPatcher.py
#!/usr/bin/python
from guizero import App, PushButton, Text, Picture
from configparser import ConfigParser
from pathlib import Path
import os, stat, sys, shutil, platform, subprocess, random, hashlib

# Check if config exists,
# if not make a new one,
# then read config file
if not Path("HoneyConfig.ini").is_file():
    shutil.copy("HoneyConfig.default.ini", "HoneyConfig.ini")
config = ConfigParser()
config.read('HoneyConfig.ini')

paths = ["mods", "template"]
for path in paths:
    if not os.path.exists(path):
        os.mkdir(path)

# set usrdir, logoskip is unused
usrdir = config.get('main', 'usrdir')
logoskip = config.getboolean('main', 'logoskip')
checksum_verif = config.getboolean('main', 'checksum')

# Make psarc executable. There's probably a better way to do this.
if platform.system().lower() != "windows":
    st = os.stat('bin/macosx/psarc')
    os.chmod('bin/macosx/psarc', st.st_mode | stat.S_IEXEC)
    st = os.stat('bin/linux/psarc')
    os.chmod('bin/linux/psarc', st.st_mode | stat.S_IEXEC)
    del st

### DEFS

# Check for mono on MacOS/Linux
def mono_check():
    if not platform.system().lower() == "windows":
        if not shutil.which("mono"):
            app.error("Error", "Could not find mono in your PATH.\nPlease install it - otherwise you won't be able to install farc mods.\n\nPlease refer to the HoneyPatcher GitHub page for more information.")
            return False

# Writes the config. Used automatically in honey_restart()
def honey_write():
    with open('HoneyConfig.ini', 'w') as configfile:
        config.write(configfile)

# Acts as a way to reload the config.
def honey_restart():
    honey_write()
    app.info("info", "HoneyPatcher will now restart to apply changes.")
    os.execv(sys.executable, ['python'] + sys.argv)

# Backs up your USRDIR.
def honey_backup():
    rom = os.path.join(usrdir, "rom.psarc")
    if not Path(rom).is_file():
        app.error("Hey!", "Your rom.psarc is missing! This indicates your game has been modified. Please install a clean copy of the game so I can make a backup.")
        return False
    else:
        if os.path.exists("BACKUP"):
            shutil.rmtree("BACKUP")
        try:
            shutil.copytree(usrdir, "BACKUP")
        except:
            app.error("Oopsies", "An error occured. This really shouldn't be possible unless permissions are messed up, but anyways...\n\n Ping Ali with this error: backup copy failed")
        else:
            app.info("Notice", "I have created a backup of your game files in a folder named \"BACKUP\" (Please don't mess with it.)")

# Restores your USRDIR backup.
def honey_restore():
    if not os.path.exists("BACKUP"):
        app.error("Error!", "You don\'t have a backup available!")
        return
    confirm_restore = app.yesno("Warning!", "This will erase your USRDIR and restore a backup. This will effectively uninstall all of your USRDIR mods and restore the game to a vanilla state.\n\nDo you want to continue?")
    if confirm_restore:
        if os.getcwd() == usrdir:
            app.error("why did you put the app here")
            return
        if os.path.exists(usrdir):
            shutil.rmtree(usrdir)
        try:
            shutil.copytree("BACKUP", usrdir)
        except:
            app.error("Oopsies", "write error here later")
        else:
            app.info("Notice", "Restore complete!")

def honey_prep():
    # Check if a backup was made, create one otherwise.
    # If creation fails for some reason, don't continue.
    if not os.path.exists("BACKUP"):
        if honey_backup() == False:
            return

    # set the path to the rom.psarc
    rom = os.path.abspath(os.path.join(usrdir, "rom.psarc"))
    # if rom.psarc doesn't exist, assume game is already prepped
    if not os.path.exists(rom):
        app.error("Error!", "rom.psarc not found! Your game is either already prepped, or corrupted!")
        return

    # set path for the psarc tool
    match platform.system().lower():
        case "linux":
            psarc_rel = os.path.join(".", "bin/linux/psarc")
            psarc = os.path.abspath(psarc_rel)
        case "darwin":
            psarc_rel = os.path.join(".", "bin/macosx/psarc")
            psarc = os.path.abspath(psarc_rel)
            app.info("NOTICE", "This hasn't been tested, there may be bugs!")
        case "windows":
            psarc_rel = os.path.join(".", "bin/win32/UnPSARC.exe")
    # TODO: test all of this shit on macos
            psarc = os.path.abspath(psarc_rel)
        case _:
            app.error("Oops!", "Your OS is not supported.")
            return
    wd = os.getcwd()
    os.chdir(usrdir)
    try:
        match platform.system().lower():
            case "windows":
                subprocess.run([psarc, rom])
            case _:
                subprocess.run([psarc, '-x', rom])
    except Exception as e:
        print(e)
        app.error("Oopsies!", "Something went wrong trying to extract rom.psarc.")
        return
    else:
        os.remove(rom)
        if platform.system().lower() == "windows":
            shutil.copytree("rom_Unpacked", "rom", dirs_exist_ok=True)
            shutil.rmtree("rom_Unpacked")
        app.info("NOTICE", "Game files have been prepped for modding!")
    finally:
        os.chdir(wd)        

# Handles whether or not to patch the eboot for logoskip
def toggle_logoskip():
    print("logoskip button pressed")
    if logoskip == True:
        config.set('main', 'logoskip', 'false')
    else:
        config.set('main', 'logoskip', 'true')
    honey_restart()

# Should be set to the USRDIR, for US this is dev_hdd0/game/NPUB30927/USRDIR
def set_directory():
    app.info("Notice", "Please select the USRDIR folder for your game.")
    directory = app.select_folder(title="Select Sonic The Fighters USRDIR", folder=usrdir)
    try:
        config.set('main', 'usrdir', directory)
    except:
        app.info("Notice", "Selection cancelled.")
    else:
        honey_restart()

# honey mob cemetery
def honey_install():

    # Check if mono is installed
    if mono_check() == False:
        return

    # Check if game is prepped
    rom = os.path.join(usrdir, "rom.psarc")
    if os.path.exists(rom):
        honey_prep()

    
    # Vars for Extraction/Compression
    rom_dir = os.path.join(usrdir, "rom")
    farclist = ["sprite/n_advstf.farc", "sprite/n_cmn.farc", "sprite/n_cmn.farc", "sprite/n_fnt.farc", "sprite/n_info.farc", "sprite/n_stf.farc", "string_array.farc", "sprite/n_advstf/texture.farc", "sprite/n_cmn/texture.farc", "sprite/n_fnt/texture.farc", "sprite/n_info/texture.farc", "sprite/n_stf/texture.farc"]
    dirlist = ["sprite/n_advstf/texture", "sprite/n_cmn/texture", "sprite/n_cmn/texture", "sprite/n_fnt/texture", "sprite/n_info/texture", "sprite/n_stf/texture", "string_array", "sprite/n_advstf", "sprite/n_cmn", "sprite/n_fnt", "sprite/n_info", "sprite/n_stf"]
    farcpack = os.path.join(".", "bin/mono/FarcPack.exe")

    # Iterate through and extract farc files
    for farc in farclist:
        farcpath = os.path.join(rom_dir, farc)
        # TODO: test this shit on windows and macos
        if platform.system().lower() == "windows":
            subprocess.run([farcpack, farcpath])
        else:
            subprocess.run(["mono", farcpack, farcpath])

    # Check through zip files
    for (dirpath, dirnames, filenames) in os.walk("mods"):
        pkglist.extend(filenames)
        break
    for pkg in pkglist:
        if pkg.startswith("."):
            pass
        elif ".zip" in pkg:
            shutil.unpack_archive(os.path.join("mods", pkg), rom_dir)

    # Compression
    for fdir in dirlist:
        dirpath = os.path.join(rom_dir, fdir)
        if platform.system().lower() == "windows":
            subprocess.run([farcpack, dirpath])
        else:
            subprocess.run(["mono", farcpack, dirpath])

    app.info("Notice", "Mod installation complete.")

    # Cleanup    
    for fdir in dirlist:
        dirpath = os.path.join(rom_dir, fdir)
        # So for some reason this fails with "texture" 
        # not found if I don't do it this way. 
        # What? Why? If I remove these lines texture remains.
        try:
            shutil.rmtree(dirpath)
        except:
            pass

def honey_unpack():

    if mono_check() == False:
        return

    rom = os.path.join(usrdir, "rom.psarc")
    if os.path.exists(rom):
        honey_prep()

    rom_dir = os.path.join(usrdir, "rom")
    farclist = ["sprite/n_advstf.farc", "sprite/n_cmn.farc", "sprite/n_cmn.farc", "sprite/n_fnt.farc", "sprite/n_info.farc", "sprite/n_stf.farc", "string_array.farc", "sprite/n_advstf/texture.farc", "sprite/n_cmn/texture.farc", "sprite/n_fnt/texture.farc", "sprite/n_info/texture.farc", "sprite/n_stf/texture.farc"]
    farcpack = os.path.join(".", "bin/mono/FarcPack.exe")

    for farc in farclist:
        farcpath = os.path.join(rom_dir, farc)
        if platform.system().lower() == "windows":
            subprocess.run([farcpack, farcpath])
        else:
            subprocess.run(["mono", farcpack, farcpath])

    if not os.path.exists("template"):
        os.makedirs("template")

    shutil.copytree(rom_dir, "template", dirs_exist_ok=True)

### GUI

app = App(title="HoneyPatcher: Arcade Stage", bg="#090F10")

logo = Picture(app, image="assets/HONEYBADGER.png")

magic_number = random.randrange(0, 99) # 1 in 100 chance of explode.png
if magic_number == 1:
    logo.image = "assets/explode.png"

select_folder_button = PushButton(app, text="Select USRDIR...", command=set_directory)
select_folder_button.text_color = "#4298f5"

# automated backups, still can be reenabled if you uncomment the below two lines
#backup_button = PushButton(app, text="Backup USRDIR...", command=honey_backup)
#backup_button.text_color = "#e7e7e7"

restore_button = PushButton(app, text="Restore USRDIR...", command=honey_restore)
restore_button.text_color = "#4298f5"

prepare_button = PushButton(app, text="Prepare USRDIR for mods...", command=honey_prep)
prepare_button.text_color = "#4298f5"

install_button = PushButton(app, text="Install all mods...", command=honey_install)
install_button.text_color = "#4298f5"

unpack_button = PushButton(app, text="Unpack game files into template...", command=honey_unpack)
unpack_button.text_color = "#4298f5"

app.display()

File: HoneyPatcher.cs
using Godot;
using Gibbed.IO;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.IO.Compression;
using System.Threading.Tasks;
using MikuMikuLibrary.Archives;
using MikuMikuLibrary.IO;
using IniParser;
using IniParser.Model;
using UnPSARC;

public partial class HoneyPatcher : Node2D
{	
	[Export]
	public AcceptDialog _acceptdialog; // Errors and whatnot
	[Export]
	public FileDialog _usrdirdialog; // Restore USRDIR button
	[Export]
	public Button _install; // Restore USRDIR button
	[Export]
	public Button _restoreusrdir; // Restore USRDIR button
	[Export]
	public Button _modsfolder; // Opens mods folder, doesn't currently work on my setup for some reason
	[Export]
	public Button _genpatches; // Generate Patches button
	[Export]
	public RichTextLabel _progress; // Progress label
	[Export]
	public LineEdit _patchname; // Name of patch
	[Export]
	public Button _patchesfolder; // Opens patches folder
	
	// Static directories
	string modsDir = ProjectSettings.GlobalizePath("user://mods");
	string workbenchDir = ProjectSettings.GlobalizePath("user://workbench");
	string backupDir = ProjectSettings.GlobalizePath("user://BACKUP");
	string honeyConfig = ProjectSettings.GlobalizePath("user://HoneyConfig.ini");
	
	string usrdir;
	string patchname = "Default";
	bool nomods = false;
	
	public override void _Ready(){	
		
		// Signal Connection
		_usrdirdialog.DirSelected += OnUsrdirDialog;
		_install.Pressed += OnInstallPressed;
		_restoreusrdir.Pressed += OnRestoreUsrdirPressed;
		_modsfolder.Pressed += OpenModsFolder;
		_genpatches.Pressed += CreatePatches;
		_patchesfolder.Pressed += OpenPatchesFolder;
		
		// Generate default config
		if(!File.Exists(honeyConfig)){
			string defaultConfig = "[main]\nlogoskip = false\nusrdir = .";
			File.WriteAllText(honeyConfig, defaultConfig);
			_progress.Text += "[I] Created default configuration file.\n";
		}
		
		string[] essentialDirs = {modsDir, workbenchDir, Path.Combine(workbenchDir, "original"), Path.Combine(workbenchDir, "modified"), Path.Combine(workbenchDir, "patches")};
		
		foreach (string h in essentialDirs){
			if (!Directory.Exists(h)){
				Directory.CreateDirectory(h);
				_progress.Text += $"[I] Created directory {h}.\n";
			}
		}
		
		// https://github.com/rickyah/ini-parser
		// MIT License
		IniData data = new FileIniDataParser().ReadFile(honeyConfig);
		usrdir = data["main"]["usrdir"];
		_progress.Text += "[I] loaded HoneyConfig.ini.\n";
		
	}

	// Signals
	private void OnUsrdirDialog(string dir){
		usrdir = Path.GetFullPath(dir); // Set usrdir for current session
		IniData data = new FileIniDataParser().ReadFile(honeyConfig); // Open config file
		data["main"]["usrdir"] = dir; // Set usrdir
		new FileIniDataParser().WriteFile(honeyConfig, data); // Write config file
		_progress.Text += "[I] Saved changes.\n";
	}
	
	// Install mods
	private void OnInstallPressed(){
		// Check for clean copy of stf w/ rom.psarc still intact
		string psarc_path = Path.Combine(usrdir, "rom.psarc");
		if (!File.Exists(psarc_path)){
			ShowError("Error", "rom.psarc could not be found. Please ensure you have a clean copy of Sonic the Fighters if this\nis your first time, or Uninstall mods before proceeding.");
			_progress.Text += "[E] rom.psarc not found.\n";
			return;
		}
		
		// Make backup if valid stf found and no backup exists
		if(!Directory.Exists(backupDir)){
			Directory.CreateDirectory(backupDir);
			_progress.Text += "[I] Created backup directory.\n";
		}
		CopyFilesRecursively(usrdir, backupDir);
		_progress.Text += "[I] Created backup.\n";
		
		// Extract rom.psarc - used UnPSARC by NoobInCoding as a base, stripped it down,
		// and turned it into a DLL. It's honestly still really bloated and could do with
		// a bit more cleanup.
		
		PsarcThing.UnpackArchiveFile(psarc_path, Path.Combine(usrdir, "rom"));
		
		_progress.Text += "[I] Extracted rom.psarc.\n";
		File.Delete(psarc_path);
		_progress.Text += "[I] Extracted and removed rom.psarc.\n";
		FarcUnpack();
		_progress.Text += "[I] Unpacked farc files.\n";
		ExtractMods();
		_progress.Text += "[I] Extracted mods.\n";
		ApplyPatches();
		_progress.Text += "[I] Applied patches.\n";
		FarcPack();
		_progress.Text += "[I] Repacked farc files.\n";
		if (nomods)
			ShowError("Success?", "No mods were found, but I extracted rom.psarc for you anyways.");
		else
			ShowError("Success!", "Mods have been installed!");
	}
	
	// Uninstall Mods
	private void OnRestoreUsrdirPressed(){
		// Check if backup exists
		if (!Directory.Exists(backupDir)){
			ShowError("Error", "No backup found.");
			_progress.Text += "[E] No backup found.\n";
			return;
		}
		
		// Clear contents of USRDIR
		try{
			// https://stackoverflow.com/questions/1288718/how-to-delete-all-files-and-folders-in-a-directory
			System.IO.DirectoryInfo di = new DirectoryInfo(usrdir);
			foreach (FileInfo file in di.GetFiles())
				file.Delete(); 
			foreach (DirectoryInfo dir in di.GetDirectories())
				dir.Delete(true); 
		}
		catch (Exception e){
			ShowError("Exception", e.ToString());
			_progress.Text += "[E] Error restoring files. (Couldn't wipe game files.)\n";
			return;
		}
		
		// Restore backup
		try{
			CopyFilesRecursively(backupDir, usrdir);
		}
		catch (Exception e){
			ShowError("Exception", e.ToString());
			_progress.Text += "[E] Error restoring files. (Couldn't copy backup.)\n";
			return;
		}
		ShowError("Success", "Files restored.");
		_progress.Text += "[I] Restored game files.\n";
	}

	private void OpenModsFolder(){
		OS.ShellOpen(modsDir);
	}
	
	private void OpenPatchesFolder(){
		OS.ShellOpen(workbenchDir);
	}
	
	private void CreatePatches(){
		if (_patchname.Text != "")
		  patchname = _patchname.Text;
		List<string> files = new List<string>();
		GD.Print("creating patches");
		string[] roms = {"rom_code1.bin", "rom_data.bin", "rom_ep.bin", "rom_pol.bin", "rom_tex.bin", "string_array_en.bin"};
		foreach (string rawhm in roms){
			if (!File.Exists(Path.Combine(workbenchDir, "original", rawhm))){
				GD.Print("original " + rawhm + " not found");
				ShowError("Error", "Original " + rawhm + "not found.");
				return;
			}
			if (File.Exists(Path.Combine(workbenchDir, "modified", rawhm))){
				GD.Print("modified " + rawhm + " found");
				files.Add(rawhm);
			}
		}
		foreach (string filename in files){
			uint count = 0;
			//uint changecount = 0;
			List<string> locations = new List<string>();
			List<byte> changes = new List<byte>();
			GD.Print("Patch Name: " + patchname);
			string patchextension = Path.GetFileNameWithoutExtension(Path.Combine(workbenchDir, "original", filename));
			GD.Print("Patch Extension: " + patchextension);
			byte[] original = File.ReadAllBytes(Path.Combine(workbenchDir, "original", filename));
			byte[] modified = File.ReadAllBytes(Path.Combine(workbenchDir, "modified", filename));
			foreach (byte b in original)
			{
				if (b != modified[count]){
					locations.Add(count.ToString());
					changes.Add(modified[count]);
					//changecount++;
				}
				count++;
			}
			GD.Print("Change Locations: " + locations.Count.ToString());
			GD.Print("Changes: " + changes.Count.ToString());
			string patch = Path.Combine(workbenchDir, "patches", patchname + "." + patchextension);
			string patchloc = patch + ".loc";
			GD.Print("Patch: " + patch);
			GD.Print("Patch Locations: " + patchloc);
			File.WriteAllBytes(patch, changes.ToArray());
			_progress.Text += $"[I] Created {patch}.\n";
			File.WriteAllLines(patchloc, locations.ToArray());
			_progress.Text += $"[I] Created {patchloc}.\n";
		}
	}
	
	public override void _Process(double delta){}

	public void ShowError(string title, string text){
		_acceptdialog.Title = title;
		_acceptdialog.DialogText = text;
		_acceptdialog.Show();
	}	
	
	// https://stackoverflow.com/questions/58744/copy-the-entire-contents-of-a-directory-in-c-sharp
	private static void CopyFilesRecursively(string sourcePath, string targetPath){
		//Now Create all of the directories
		foreach (string dirPath in Directory.GetDirectories(sourcePath, "*", SearchOption.AllDirectories)){
			Directory.CreateDirectory(dirPath.Replace(sourcePath, targetPath));
		}
		//Copy all the files & Replaces any files with the same name
		foreach (string newPath in Directory.GetFiles(sourcePath, "*.*",SearchOption.AllDirectories)){
			File.Copy(newPath, newPath.Replace(sourcePath, targetPath), true);
		}
	}
	
	/* The code for the next two functions was almost entirely ripped from
	Skyth's FarcPack utility, and uses MikuMikuLibrary to unpack/repack game 
	files. It is licensed under the MIT license. Please show him your support.
	https://github.com/blueskythlikesclouds/MikuMikuLibrary */
	
	private void FarcUnpack(){
		string[] farcs =   {"sprite/n_advstf.farc", 
							"sprite/n_cmn.farc", 
							"sprite/n_cmn.farc", 
							"sprite/n_fnt.farc", 
							"sprite/n_info.farc", 
							"sprite/n_stf.farc", 
							"string_array.farc", 
							"sprite/n_advstf/texture.farc", 
							"sprite/n_cmn/texture.farc", 
							"sprite/n_fnt/texture.farc", 
							"sprite/n_info/texture.farc", 
							"sprite/n_stf/texture.farc"};
	
		foreach (string farc in farcs )
		{
			// Set source and destination filename
			string sourceFileName = Path.Combine(usrdir, "rom", farc);
			string destinationFileName = Path.ChangeExtension(sourceFileName, null);
					
			using (var stream = File.OpenRead(sourceFileName))
			{
				var farcArchive = BinaryFile.Load<FarcArchive>(stream);
				
				Directory.CreateDirectory(destinationFileName);
				
				foreach (string fileName in farcArchive)
				{
					using (var destination = File.Create(Path.Combine(destinationFileName, fileName)))
					using (var source = farcArchive.Open(fileName, EntryStreamMode.OriginalStream))
					source.CopyTo(destination);
				}
			}
		}
	}
	
	private void FarcPack(){
		string[] dirlist = {"sprite/n_advstf/texture", 
							"sprite/n_cmn/texture", 
							"sprite/n_cmn/texture", 
							"sprite/n_fnt/texture", 
							"sprite/n_info/texture", 
							"sprite/n_stf/texture", 
							"string_array", 
							"sprite/n_advstf", 
							"sprite/n_cmn", 
							"sprite/n_fnt", 
							"sprite/n_info", 
							"sprite/n_stf"};
	
		foreach (string dir in dirlist)
		{
			// Set source and destination file name
			string sourceFileName = Path.GetFullPath(Path.Combine(usrdir, "rom", dir));
			string destinationFileName = Path.ChangeExtension(sourceFileName, "farc");;

			// These arguments should never change, as they seem to work fine with STF
			bool compress = false;
			int alignment = 16;
			
			// Modified by me, otherwise it throws access errors if you don't use "using"
			using (var farcArchive = new FarcArchive { IsCompressed = compress, Alignment = alignment }){
			
			if (File.GetAttributes(sourceFileName).HasFlag(FileAttributes.Directory))
			{
				foreach (string filePath in Directory.EnumerateFiles(sourceFileName))
				farcArchive.Add(Path.GetFileName(filePath), filePath);
			}
			else
			{
				farcArchive.Add(Path.GetFileName(sourceFileName), sourceFileName);
			}
			farcArchive.Save(destinationFileName);
			}
		}
	}
	
	private void ExtractMods(){
		string[] files = Directory.GetFiles(modsDir);
		Array.Sort(files);
		if (files.Length == 0){
			nomods = true;
			return;
		}
		else{
			nomods = false;
		}
		
		foreach (string mod in files)
		{
			string modpath = mod;
			string romdir = Path.Combine(usrdir, "rom");
			string stf_rom = Path.Combine(romdir, "stf_rom");
			if (Path.GetExtension(modpath) == ".zip")
				ZipFile.ExtractToDirectory(modpath, romdir, true);
		}
	}
	
	private void ApplyPatches(){
		// Get list of files in rom folder
		string[] files = Directory.GetFiles(Path.Combine(usrdir, "rom"));
		// Apply in alphabetical order
		Array.Sort(files);
		foreach (string mod in files)
		{
			string modpath = mod; // patch
			string romdir = Path.Combine(usrdir, "rom");
			string stf_rom = Path.Combine(romdir, "stf_rom");
			string patchdest; // file to be patched
			switch(Path.GetExtension(modpath))
			{
				// Check the file extension, which should be the name of the file you want to patch
				case ".rom_code1":
					patchdest = Path.Combine(stf_rom, "rom_code1.bin");
					break;
				case ".rom_data":
					patchdest = Path.Combine(stf_rom, "rom_data.bin");
					break;
				case ".rom_ep":
					patchdest = Path.Combine(stf_rom, "rom_ep.bin");
					break;
				case ".rom_pol":
					patchdest = Path.Combine(stf_rom, "rom_pol.bin");
					break;
				case ".rom_tex":
					patchdest = Path.Combine(stf_rom, "rom_tex.bin");
					break;
				// At some point we'll handle these with actual XML extraction/injection!
				// For now, this will do.
				case ".string_array_en":
					patchdest = Path.Combine(romdir, "string_array", "string_array_en.bin");
					break;
				case ".string_array_jp":
					patchdest = Path.Combine(romdir, "string_array", "string_array_jp.bin");
					break;
				default:
					patchdest = null;
					break;
			}
			if (patchdest != null){
				// modpath = patch
				// patchdest = file to be patched
				try{
					byte[] changes = File.ReadAllBytes(modpath);
					string[] locations = File.ReadAllLines(modpath+".loc");
					uint inc = 0;
					using (FileStream fs = File.Open(patchdest, FileMode.Open, System.IO.FileAccess.ReadWrite, FileShare.ReadWrite)){
						foreach (string i in locations){
							long loc = Int64.Parse(i);
							fs.Seek(loc, SeekOrigin.Begin);
							// GD.Print(loc.ToString());
							fs.WriteByte(changes[inc]);
							inc++;
						}
					}
				}
				catch(Exception e){
					// Need to write error handling here later, this will do for now
					GD.Print(e.ToString());
				}
			}
		}
	}
}
